src/lib/core/datetime.c

```diff
@@ -60,13 +60,61 @@ datetime_to_tm(const struct datetime *date, struct tnt_tm *tm)
 
 size_t
 datetime_strftime(const struct datetime *date, char *buf, size_t len,
-		      const char *fmt)
+		  const char *fmt)
 {
+	assert(date != NULL);
 	struct tnt_tm tm;
 	datetime_to_tm(date, &tm);
 	return tnt_strftime(buf, len, fmt, &tm);
 }
 
+void
+tm_to_datetime(struct tnt_tm *tm, struct datetime *date)
+{
+	assert(tm != NULL);
+	assert(date != NULL);
+	int year = tm->tm_year;
+	int mon = tm->tm_mon;
+	int mday = tm->tm_mday;
+	int yday = tm->tm_yday;
+	int wday = tm->tm_wday;
+	dt_t dt = 0;
+
+	if ((year | mon | mday) == 0) {
+		if (yday != 0) {
+			dt = yday - 1 + DT_EPOCH_1970_OFFSET;
+		} else if (wday != 0) {
+			/* 1970-01-01 was Thursday */
+			dt = ((wday - 4) % 7) + DT_EPOCH_1970_OFFSET;
+		}
+	} else {
+		assert(mday >= 0 && mday < 32);
+		assert(mon >= 0 && mon <= 11);
+		dt = dt_from_ymd(year + 1900, mon + 1, mday);
```

Comment:
Should be dt_from_ymd_checked?
Otherwise the lower date boundary simply wraps around.
```
tarantool> date.parse("-5879610-06-22", {format="%Y-%m-%d"})
---
- -5879610-06-22T00:00:00Z
- 14
...

tarantool> date.parse("-5879610-06-21", {format="%Y-%m-%d"})
---
- 5879611-07-11T00:00:00Z
- 14
...


```

Reply:
Thanks! Fixed:
```
tarantool> date.parse("-5879610-06-22", {format="%Y-%m-%d"})
---
- -5879610-06-22T00:00:00Z
- 14
...

tarantool> date.parse("-5879610-06-21", {format="%Y-%m-%d"})
---
- error: 'builtin/datetime.lua:930: could not parse ''-5879610-06-21'' using ''%Y-%m-%d''
    format'
...
```

Reply:
I had to introduce changes both in Tarantool and c-dt.

Here are Tarantool changes:
```diff
diff --git a/extra/exports b/extra/exports
index be61fb7c1..efbe17eef 100644
--- a/extra/exports
+++ b/extra/exports
@@ -430,6 +430,7 @@ tnt_dt_dow
 tnt_dt_doy
 tnt_dt_from_rdn
 tnt_dt_from_ymd
+tnt_dt_from_ymd_checked
 tnt_dt_month
 tnt_dt_parse_iso_date
 tnt_dt_parse_iso_time
diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index 698e152ec..9c1d15d50 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -9,6 +9,7 @@
 #include <string.h>
 #include <time.h>
 
+#define DT_PARSE_ISO_TNT
 #include "c-dt/dt.h"
 #include "datetime.h"
 #include "trivia/util.h"
@@ -68,7 +69,7 @@ datetime_strftime(const struct datetime *date, char *buf, size_t len,
 	return tnt_strftime(buf, len, fmt, &tm);
 }
 
-void
+bool
 tm_to_datetime(struct tnt_tm *tm, struct datetime *date)
 {
 	assert(tm != NULL);
@@ -88,9 +89,12 @@ tm_to_datetime(struct tnt_tm *tm, struct datetime *date)
 			dt = ((wday - 4) % 7) + DT_EPOCH_1970_OFFSET;
 		}
 	} else {
-		assert(mday >= 0 && mday < 32);
+		if (mday == 0)
+			mday = 1;
+		assert(mday >= 1 && mday <= 31);
 		assert(mon >= 0 && mon <= 11);
-		dt = dt_from_ymd(year + 1900, mon + 1, mday);
+		if (dt_from_ymd_checked(year + 1900, mon + 1, mday, &dt) == false)
+			return false;
 	}
 	int64_t local_secs =
 		(int64_t)dt * SECS_PER_DAY - SECS_EPOCH_1970_OFFSET;
@@ -99,6 +103,7 @@ tm_to_datetime(struct tnt_tm *tm, struct datetime *date)
 	date->nsec = tm->tm_nsec;
 	date->tzindex = 0;
 	date->tzoffset = tm->tm_gmtoff / 60;
+	return true;
 }
 
 size_t
@@ -111,7 +116,8 @@ datetime_strptime(struct datetime *date, const char *buf, const char *fmt)
 	char * ret = tnt_strptime(buf, fmt, &t);
 	if (ret == NULL)
 		return 0;
-	tm_to_datetime(&t, date);
+	if (tm_to_datetime(&t, date) == false)
+		return 0;
 	return ret - buf;
 }
 
diff --git a/third_party/c-dt b/third_party/c-dt
--- a/third_party/c-dt
+++ b/third_party/c-dt
@@ -1 +1 @@
-Subproject commit 5afed1063eb8b4dc0a12f7e753456862ca4de38b
+Subproject commit 5afed1063eb8b4dc0a12f7e753456862ca4de38b-dirty
```

And here are `c-dt` changes (collapsed)...
```diff
diff --git a/dt_rename.h b/dt_rename.h
index da21fc6..5213bfd 100644
--- a/dt_rename.h
+++ b/dt_rename.h
@@ -55,6 +55,9 @@
 #  define dt_from_ymd DT_NAME(DT_NAMESPACE, dt_from_ymd)
 #  define dt_from_yqd DT_NAME(DT_NAMESPACE, dt_from_yqd)
 #  define dt_from_ywd DT_NAME(DT_NAMESPACE, dt_from_ywd)
+#  ifdef DT_PARSE_ISO_TNT
+#    define dt_from_ymd_checked DT_NAME(DT_NAMESPACE, dt_from_ymd_checked)
+#  endif
 #  define dt_leap_year DT_NAME(DT_NAMESPACE, dt_leap_year)
 #  define dt_month DT_NAME(DT_NAMESPACE, dt_month)
 #  define dt_parse_iso_date DT_NAME(DT_NAMESPACE, dt_parse_iso_date)
```